cayenne/src/main/java/org/apache/cayenne/access/translator/select/OrderingAbstractStage.java

```diff
@@ -0,0 +1,107 @@
+import org.apache.cayenne.access.sqlbuilder.sqltree.ColumnNode;
+import org.apache.cayenne.access.sqlbuilder.sqltree.FunctionNode;
+import org.apache.cayenne.access.sqlbuilder.sqltree.Node;
+import org.apache.cayenne.access.sqlbuilder.sqltree.NodeType;
+import org.apache.cayenne.access.sqlbuilder.sqltree.SelectResultNode;
+import org.apache.cayenne.access.sqlbuilder.sqltree.SimpleNodeTreeVisitor;
+import org.apache.cayenne.exp.Expression;
+import org.apache.cayenne.exp.parser.ASTAggregateFunctionCall;
+import org.apache.cayenne.map.DbAttribute;
+import org.apache.cayenne.query.Ordering;
+
+abstract class OrderingAbstractStage implements TranslationStage {
+
+    protected void processOrdering(QualifierTranslator qualifierTranslator, TranslatorContext context, Ordering ordering) {
+        Expression orderExp = ordering.getSortSpec();
+        NodeBuilder nodeBuilder = node(qualifierTranslator.translate(orderExp));
+
+        if(ordering.isCaseInsensitive()) {
+            nodeBuilder = function("UPPER", nodeBuilder);
+        }
+
+        // If query is DISTINCT or GROUPING then we need to add the order column as a result column
+        if (orderColumnAbsent(context, nodeBuilder)) {
+            // deepCopy as some DB expect exactly the same expression in select and in ordering
+            ResultNodeDescriptor descriptor = context.addResultNode(nodeBuilder.build().deepCopy());
+            if(orderExp instanceof ASTAggregateFunctionCall) {
+                descriptor.setAggregate(true);
+            }
+        }
+    }
+
+    private DbAttribute getOrderDbAttribute(Node translatedOrderNode)
+    {
+        DbAttribute[] orderDbAttribute = {null};
+        translatedOrderNode.visit(new SimpleNodeTreeVisitor() {
+            @Override
+            public boolean onNodeStart(Node node) {
+                if (node.getType() == NodeType.COLUMN) {
+                    orderDbAttribute[0] = ((ColumnNode) node).getAttribute();
+                    return false;
+                }
+                return true;
+            }
+        });
+        return orderDbAttribute[0];
+    }
+
+    private boolean orderColumnAbsent(TranslatorContext context, NodeBuilder nodeBuilder)
+    {
+        var orderDbAttribute = getOrderDbAttribute(nodeBuilder.build());
+        if (orderDbAttribute == null) return false; // Alias ?
+
+        var orderEntity = orderDbAttribute.getEntity().getName();
+        var orderColumn = orderDbAttribute.getName();
+
+        Predicate<DbAttribute> columnAndEntity = dba -> dba != null
+                                    && orderColumn.equals(dba.getName())
+                                    && orderEntity.equals(dba.getEntity().getName());
+
+        var orderStr = getSqlString(order(nodeBuilder));
+
+        return context.getResultNodeList().stream()
+            .filter( result -> columnAndEntity.test(result.getDbAttribute()) )
+            .noneMatch( result -> getSqlString(node(result.getNode())).startsWith(orderStr) );
```

Comment:
Note that in `getSqlString` a trailing space is appended in order to prevent false positives like this.

Reply:
Yeah, and `DbAttribute` filter should help too. Probably we could still get wrong results with complex orderings. Like `SELECT column FROM table ORDER BY column + 1`

Reply:
So for `SELECT DISTINCT column FROM table ORDER BY column + 1` we would get:
```
"column ".startsWith("column + 1 ") -> false
```
and so produces the following SQL: `SELECT DISTINCT column, column + 1 FROM table ORDER BY column + 1`
which is correct isn't ?

Reply:
Yes this case should be ok. But I suppose the opposite case will be incorrect. I.e. `SELECT DISTINCT column + 1 FROM table ORDER BY column` will result in `"column + 1 ".startsWith("column ") -> true`

Reply:
But reversing your example will fail for: `SELECT DISTINCT column + 1 FROM table ORDER BY column` we would get:
```
"column + 1 ".startsWith("column ") -> true
```
and so produces the following SQL: `SELECT DISTINCT column + 1 FROM table ORDER BY column`
which fails !

Reply:
Actually I'm wrong wrt my last example because Cayenne wraps the expression in brackets like so:
`SELECT DISTINCT ( column + 1 ) FROM table ORDER BY column` so we then get:
```
"( column + 1 ) ".startsWith("column ") -> false
```
and the resulting SQL will be: `SELECT DISTINCT ( column + 1 ), column FROM table ORDER BY column`

Reply:
Ok, so it's a bit fragile but should hold for now.

Reply:
> Ok, so it's a bit fragile but should hold for now.

I agree on it being fragile and have given it some thought. I think we have 3 options:
1.  Change it to **_equals_** instead. This will be more robust at the expense of duplicating only aliased columns.
2.  I can try and do something with Node.visit to somehow compare the order Node with the result column nodes.
3.  Or add an abstract method "matches/equals( Node )" and method "deepEquals( Node )" to Node.

I'm willing to do either of these, which do you prefer ?

Reply:
I think I'll give option 2 a shot first then. It'll take me a couple of days though (maybe next week) as I've got some other commitments .....

Reply:
@stariy95 I'm ready to submit a new PR for the Node.visit pattern but need some guidance on how to go about it ?
Should I create a new Jira RFE issue and submit against that, or just do an ordinary PR ?

Reply:
@Jugen If it's more than a few lines of code, then it's better to open a new Jira issue.

Reply:
@stariy95 Have submitted PR #610 (ready for review) against [CAY-2847](https://issues.apache.org/jira/browse/CAY-2847) to address this.

Reply:
Thank you! I've seen it but unfortunately didn't have time to review it. Hope will get some free time to check it out in the next few days.